photon-client/src/main/scala/com/linkedin/photon/ml/data/GameConverters.scala

```diff
@@ -19,55 +19,47 @@ import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.{DataFrame, Row}
 
+import com.linkedin.photon.ml.InputColumnsNames
 import com.linkedin.photon.ml.util.VectorUtils
 
 /**
  * A collection of utility functions for converting to and from GAME datasets.
  */
 object GameConverters {
 
-  /**
-   * Standard field names
-   */
-  object FieldNames {
-    val RESPONSE: String = "response"
-    val OFFSET: String = "offset"
-    val WEIGHT: String = "weight"
-    val UID: String = "uid"
-    val META_DATA_MAP: String = "metadataMap"
-  }
-
   // Column name for the synthesized unique id column
-  val UNIQUE_ID_COLUMN_NAME = "___photon:uniqueId___"
+  private val UNIQUE_ID_COLUMN_NAME = "___photon:uniqueId___"
 
   /**
    * Converts a DataFrame into an [[RDD]] of type [[GameDatum]].
    *
+   * @note We "decode" the Map of column names into an Array[String] which we broadcast for performance.
+   *
    * @param data The source DataFrame
    * @param featureShards A set of feature shard ids
    * @param idTypeSet A set of id types expected to be found in the row
    * @param isResponseRequired Whether the response variable is expected to be found in the row. For example, if GAME
    *   data set to be parsed is used for model training, then the response variable is expected to be found in row. If
    *   the GAME data set is used for scoring, then we don't expect to find response.
+   * @param inputColumnsNames User-supplied input column names to read the input data
    * @return The [[RDD]] of type [[GameDatum]]
    */
   protected[ml] def getGameDataSetFromDataFrame(
-    data: DataFrame,
-    featureShards: Set[String],
-    idTypeSet: Set[String],
-    isResponseRequired: Boolean): RDD[(Long, GameDatum)] = {
+      data: DataFrame,
+      featureShards: Set[String],
+      idTypeSet: Set[String],
+      isResponseRequired: Boolean,
+      inputColumnsNames: Map[InputColumnsNames.Value, String]): RDD[(Long, GameDatum)] = {
 
-    // Add unique id
     val recordsWithUniqueId = data.withColumn(UNIQUE_ID_COLUMN_NAME, monotonicallyIncreasingId)
+    // Convert the (possibly partially populated) map of user-specified input columns names to a smaller array of
+    // strings, but follow the order of the fields in InputColumnsNames so that we can use InputColumnsNames field
+    // indexes to access user-defined column names directly
+    val inputColumnsNamesArray = inputColumnsNames.toArray.map { case (k, v) => (k.id, v) }.sortBy(_._1).map(_._2)
```

Comment:
Well - I think it's too complicated with an array, and we don't know if the gain is there - the idea was that indexing through an array was faster than a map, since we are going to look up names a lot, for each row of the data. But we can wait and see on performance. Alex is adamant that he wants a broadcast, so I guess I'll do a broadcast or a `Map[String, String]` in the end.